### ISSUE TITLE
[Bug] Update snapshot logic to use dbt.hash instead of hardcoded md5

### ISSUE THREAD

**Issue by graciegoheen** [dbt-labs/dbt-core#7466]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Currently, `dbt_scd_id` is built with a hardcoded [md5](https://github.com/dbt-labs/dbt-core/blob/828d723512fced809c63e369a82c7eb570a74986/core/dbt/include/global_project/macros/materializations/snapshots/strategies.sql#L43) 

### Expected Behavior

Instead, `dbt_scd_id` should be built use `dbt.hash` to remove hard-coded md5

### Steps To Reproduce

Run a dbt snapshot

### Environment

```markdown
- OS:
- Python:
- dbt:
```

**Comment by jtcohen6** [comment]
@graciegoheen Are you thinking about other data platforms that don't support `md5`, or support it slightly differently?

IIRC, BigQuery is one such example. Looks like we've reimplemented this macro [over there](https://github.com/dbt-labs/dbt-bigquery/blob/cf794aaa44ad7b426fe0dec125c3d15df741ad73/dbt/include/bigquery/macros/materializations/snapshot.sql#L1-L6):
```sql
{% macro bigquery__snapshot_hash_arguments(args) -%}
  to_hex(md5(concat({%- for arg in args -%}
    coalesce(cast({{ arg }} as string), ''){% if not loop.last %}, '|',{% endif -%}
  {%- endfor -%}
  )))
{%- endmacro %}
```
When we could just be using the cross-db `hash()` utility, and saving ourselves the trouble!

**Comment by graciegoheen** [comment]
That's right. For your BigQuery example, we already have a [hash override for BigQuery](https://github.com/dbt-labs/dbt-bigquery/blob/87a68e55ea690fe805b004012e9701752b1ddfe5/dbt/include/bigquery/macros/utils/hash.sql), so also having the `snapshot_hash_arguements` override feels redundant:
```
{% macro bigquery__hash(field) -%}
    to_hex({{dbt.default__hash(field)}})
{%- endmacro %}
```
The other different here for BigQuery's `snapshot_hash_arguements` is that it requires the elements to be cast to string (instead of [varchar in the default](https://github.com/dbt-labs/dbt-core/blob/828d723512fced809c63e369a82c7eb570a74986/core/dbt/include/global_project/macros/materializations/snapshots/strategies.sql#L44)), but couldn't that instead be handled by using `dbt.type_string`?

It looks like Spark does something similar for [`snapshot_hash_arguments`](https://github.com/dbt-labs/dbt-spark/blob/8987e66e6e2ca721c5002bcefcfd2447e284ceaf/dbt/include/spark/macros/materializations/snapshot.sql#L3).

**Comment by dbeatty10** [comment]
Consider me a fan of cross-database implementations! 💪 

If a more cross-database implementation of `default__snapshot_hash_arguments` is added to dbt-core, it would probably involve [`dbt.type_string`](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#type_string), [`dbt.concat`](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#concat), and [`dbt.hash`](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#hash) as mentioned by @graciegoheen.

The implementation might feel a bit complicated due to `dbt.concat` taking a Jinja array of strings as its argument. I'm guessing that list would need to be built up prior to passing it into `dbt.concat`, which might make the logic harder to read and understand.

But would be nice if this worked out so we could remove more hard-coded SQL function names from dbt-core.
